File: Daily/PY/Leetcode1711-大餐计数.py
from typing import List
from collections import defaultdict

class Solution:
    def countPairs(self, deliciousness: List[int]) -> int:
        res = 0
        MOD = int(1e9 + 7)
        mp = defaultdict(int)
        m = max(deliciousness) * 2
        for d in deliciousness:
            i = 1
            while i <= m:
                pair = mp[i - d]
                res = (pair + res) % MOD
                i <<= 1
            mp[d] += 1
        return res

File: Leetcode/sbgo/Leetcode1711-大餐计数.go
package sbgo

func countPairs(deliciousness []int) int {
	mp := map[int]int{}
	res := 0
	m := -1
	for _, val := range deliciousness {
		if m < val {
			m = val
		}
	}
	m = m * 2
	for _, d := range deliciousness {
		for i := 1; i <= m; i <<= 1 {
			res = (res + mp[i-d]) % (1e9 + 7)
		}
		mp[d]++
	}
	return res
}
